### PR TITLE
Miri read_discriminant: return a scalar instead of raw underlying bytes

### PR DESCRIPTION
**Pull request by RalfJung**
r? @oli-obk @eddyb

### CODE REVIEW

File: src/librustc_mir/interpret/operand.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/operand.rs:717]: This one particularly confused me -- we should resolve that question before landing.

**Comment by RalfJung** [comment]
This makes the generator-related Miri tests fail:
```
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `Size { raw: 4 }`,
 right: `Size { raw: 8 }`: Size mismatch when writing bits', /home/r/src/rust/rustc.2/src/libstd/macros.rs:16:9
```
I'll need to investigate.

File: src/librustc_mir/interpret/operand.rs
Comment (eddyb) [inline src/librustc_mir/interpret/operand.rs:594]: This doesn't feel right, I believe the `u8` type is only used when the value is always `0`.

File: src/librustc_mir/interpret/operand.rs
Comment (eddyb) [inline src/librustc_mir/interpret/operand.rs:606]: "storing the discriminant" could be something more like "*encoding* the discriminant as a tag/niche".

That might make it clearer why the type differs: the only reason to have a simple encoding is cheap decoding, it could be any arbitrary mapping as long as you have enough values for every possible discriminant to be represented.

File: src/librustc_mir/interpret/operand.rs
Comment (eddyb) [inline src/librustc_mir/interpret/operand.rs:609]: Not sure what this refers to. Is this a tag/niche-sized integer? So for e.g. a pointer niche, it would be a pointer-sized integer?

File: src/librustc_mir/interpret/operand.rs
Comment (eddyb) [inline src/librustc_mir/interpret/operand.rs:668]: Discriminant and variant index are the same, i.e. niche layout isn't used if you use custom discriminants and assign anything to them other than their respective variant index (or any custom discriminants at all - note that this is still an unstable feature).

File: src/librustc_mir/interpret/operand.rs
Comment (eddyb) [reply src/librustc_mir/interpret/operand.rs:715]: For the niche layout, the variant indices have to not overflow host `u32` *nor* the target (type-checking) discriminant type (`isize` by default but I guess it could be `u8`/`i8`).

So you can add an assert that it fits as a positive number in the discriminant type, but no cast should be necessary.

**Review by eddyb — APPROVED** [review]
LGTM modulo comments (@oli-obk could take a look at the more miri-specific parts of this, which I'm not too familiar with)

File: src/librustc_mir/interpret/operand.rs
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs:594]: So how do I get the right type then?

File: src/librustc_mir/interpret/operand.rs
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs:606]: > "encoding the discriminant as a tag/niche".

But the layout of the tag/niche is `discr_layout`, right? The second item in my list? And then `discr_ty` is the high-level thing?

Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs:606]: > But the layout of the tag/niche is discr_layout, right? 

If this is correct I'll rename the variables to `tag_layout` or so.

File: src/librustc_mir/interpret/operand.rs
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs:609]: This is the layout stored in `Variants::Multiple::discr`. So I think the answer to your questions is "yes".

File: src/librustc_mir/interpret/operand.rs
Comment (eddyb) [reply src/librustc_mir/interpret/operand.rs:601]: The type is right, I should've just said that you can probably assert the index is `0` here.

File: src/librustc_mir/interpret/operand.rs
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs]: `Scalar::from_u32` already has such an assert (at least on debug builds).

File: src/librustc_mir/interpret/operand.rs
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs]: Did that, and also opened https://github.com/rust-lang/rust/pull/72497 for the general terminology cleanup.

File: src/librustc_mir/interpret/operand.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/operand.rs:640]: So, something is wrong here. I have a generator where `discr_ty` here returns a 32bit type, but then the MIR `Discriminant` statement is used to put this into a 64bit type. IOW, the MIR is ill-formed, as far as I can tell.

[Here's the code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=75c57a7439eb29b9f9ba346ab5d565ca), and here the relevant MIR snippet:
```
fn basic::{{closure}}#0(_1: std::pin::Pin<&mut [generator@src/main.rs:37:15: 37:25 {usize, ()}]>, _2: ()) -> std::ops::GeneratorState<usize, ()> {
    let mut _0: std::ops::GeneratorState<usize, ()>; // return place in scope 0 at src/main.rs:37:15: 37:25
    let mut _3: ();                      // in scope 0 at src/main.rs:37:18: 37:18
    let mut _4: isize;                   // in scope 0 at src/main.rs:37:15: 37:25

    bb0: {
        _4 = discriminant((*(_1.0: &mut [generator@src/main.rs:37:15: 37:25 {usize, ()}]))); // scope 0 at src/main.rs:37:15: 37:25
        switchInt(move _4) -> [0u32: bb1, 1u32: bb3, 3u32: bb2, otherwise: bb4]; // scope 0 at src/main.rs:37:15: 37:25
    }
```
Notice `_4` is ptr-sized, but (if I interpret everything correctly), `discr_ty` for that generator is `u32` -- that's what my logs say, anyway:
```
[TRACE rustc_mir::interpret::operand] tag value: 0x00
[TRACE rustc_mir::interpret::operand] discriminant type: u32
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `Size { raw: 4 }`,
 right: `Size { raw: 8 }`: Size mismatch when writing bits', /home/r/src/rust/rustc.2/src/libstd/macros.rs:16:9
```

Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs]: https://github.com/rust-lang/rust/pull/72502 fixes this.

**Comment by RalfJung** [comment]
This PR now also includes a `discriminant_ty` helper method to determine the discriminant type for any type. This is shared between `Rvalue` type computation and Miri. And indeed it fixes a bug in the former which uses `isize` as discriminant type for non-enum ADTs (instead of `u8`).

**Comment by RalfJung** [comment]
Rebased as my cast-cleanup-PR also included a commit from this one.

**Comment by RalfJung** [comment]
Failed in https://github.com/rust-lang/rust/pull/72737#issuecomment-636108388
@bors r-

**Comment by RalfJung** [comment]
Fixes type sanity check assertions.
@bors r=oli-obk,eddyb

**Pull request merged**
